util/chplenv/third_party_utils.py

```diff
@@ -137,15 +137,29 @@ def pkgconfig_get_system_version(pkg):
 #
 # This returns the default link args for the given third-party package.
 #
-def default_get_link_args(pkg, ucp='', libs=[]):
+def default_get_compile_args(pkg, ucp=''):
+    if ucp == '':
+        ucp = default_uniq_cfg_path()
+    inc_dir = os.path.join(get_cfg_install_path(pkg, ucp), 'include')
+    return ['-I' + inc_dir]
+
+
+#
+# This returns the default link args for the given third-party package.
+#
+def default_get_link_args(pkg, ucp='', libs=[], add_L_opt=False):
     if ucp == '':
         ucp = default_uniq_cfg_path()
     if libs == []:
         libs = [ 'lib' + pkg + '.la' ]
     all_args = []
+    lib_dir = os.path.join(get_cfg_install_path(pkg, ucp), 'lib')
+    if add_L_opt:
+        all_args.append('-L' + lib_dir)
+        all_args.append('-Wl,-rpath,' + lib_dir)
```

Comment:
Can you explain the rpath adding here? For other third-party libs I think the rpath logic is in runtime makefiles. I'm fine with this being here, just want to understand the distinction

Reply:
Right, for the other third-party stuff this is in the Makefiles. Currently all or nearly all of the third-party support needed for building the runtime is in the Makefiles, while the third-party support for linking user program is partly in the scripts and partly in the Makefiles. I think it resulted from our realizing we couldn't do everything we wanted in the Makefiles, moving some of the processing into the scripts, but stopping progress on that as soon as things worked well enough.  (And I'm not placing blame in any sense here, it's a natural consequence of how we've worked and I'm as or more responsible than anyone else.) That seems an unfortunate complication to me, and I'd really like to get it all moved into the scripts.  (Where by "all" I mean constructing the include and library paths and building the compilation and linking options.)  So this is a step down that path. There's a note about this in the log entry for the first commit (0940970), though I probably should also have mentioned it in the 'scripts' commit:

> Note that in a departure from previous practice for most other packages, the Makefile.include here does not define LIBFABRIC_INCLUDE_DIR, LIBFABRIC_LIB_DIR, or LIBFABRIC_BIN_DIR.  These and the corresponding Makefile*libfabric-libfabric in runtime/make for the runtime build and runtime/etc for user program builds are not needed because the libfabric specific scripts in util/chplenv put the libfabric-related options in CHPL_THIRD_PARTY_COMPILE_ARGS and CHPL_THIRD_PARTY_LINK_ARGS instead. I'm inclined to think this is a cleaner solution for third-party compile and link args, so I'm trying it out here.  It's easy to change in the future if we decide to switch libfabric to the existing practice or update the other third-party packages to the new style, either way.

Reply:
Makes sense. Is there an open issue for migrating the rest of the rpath (and whatever else) for other 3rd party packages to the chplenv scripts?

Reply:
I thought I had one, but no.  I'll make one.

Reply:
Cray/chapel-private#994